### PR TITLE
Mesh optimizer and stuff

### PR DESCRIPTION
**Pull request by Crisspl**

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
will give it a look real soon :)

File: source/Irrlicht/CMeshManipulator.cpp [hunk 1]
```diff
@@ -1085,10 +1085,12 @@ ICPUMeshBuffer* CMeshManipulator::createMeshBufferWelded(ICPUMeshBuffer *inbuffe
 
 ICPUMeshBuffer* CMeshManipulator::createOptimizedMeshBuffer(const ICPUMeshBuffer* _inbuffer, const SErrorMetric* _requantErrMetric) const
 {
-	if (!_inbuffer || !_inbuffer->getMeshDataAndFormat())
-		return const_cast<ICPUMeshBuffer*>(_inbuffer);
+	if (!_inbuffer)
+		return NULL;
 
 	ICPUMeshBuffer* outbuffer = createMeshBufferDuplicate(_inbuffer);
+	if (!outbuffer->getMeshDataAndFormat())
```
Comment (devshgraphicsprogramming) [inline source/Irrlicht/CMeshManipulator.cpp]: whats going on here?

why would createMeshBufferDuplicate give a null meshdataandformat if _inbuffer has a non null one ???

File: include/IMesh.h [hunk 2]
```diff
@@ -92,10 +95,18 @@ namespace scene
...
+			@param _stackSize Size of stack memory pointed by _stackPtr.
+			@returns Pointer to memory on which blob was written.
+		*/
 		virtual void* serializeToBlob(void* _stackPtr = NULL, const size_t& _stackSize = 0) const
 		{
 			return core::CorrespondingBlobTypeFor<ICPUMesh>::type::createAndTryOnStack(this, _stackPtr, _stackSize);
 		}
+
+		//! Adds a mesh buffer.
+		virtual void addMeshBuffer(ICPUMeshBuffer* buf) = 0;
```
Comment (devshgraphicsprogramming) [inline include/IMesh.h]: There's an SGPUMesh for such operations on an IGPUMesh

should there be an SCPUMesh derived from ICPUMesh too?
(if one wants to add and remove meshbuffers)

File: include/IMesh.h [hunk 3]
```diff
@@ -42,6 +42,9 @@ namespace scene
 		virtual ~IMesh() {}
 	public:
 
+		//! Resets bounding box and removes all meshbuffers (calls drop() on each).
+		virtual void clearMeshBuffers() = 0;
```
Comment (devshgraphicsprogramming) [inline include/IMesh.h]: don't add mutability to ICPUMesh API, if you have code that requires mutability and manipulation then create an SCPUMesh (or skinned mesh) and cast it when you're done

File: source/Irrlicht/CSkinnedMesh.cpp [hunk 4]
```diff
@@ -447,12 +453,12 @@ scene::SCPUSkinMeshBuffer *CCPUSkinnedMesh::addMeshBuffer()
 }
 
 
-void CCPUSkinnedMesh::addMeshBuffer(SCPUSkinMeshBuffer* buf)
+void CCPUSkinnedMesh::addMeshBuffer(ICPUMeshBuffer* buf)
```
Comment (devshgraphicsprogramming) [inline source/Irrlicht/CSkinnedMesh.cpp]: CCPUSkinnedMesh should ONLY contain SCPUSkinMeshBuffers!!!

Comment (Crisspl) [reply source/Irrlicht/CSkinnedMesh.cpp]: And there's appropriate dynamic_cast

File: source/Irrlicht/CMeshManipulator.cpp (hunk 1 above)
Comment (Crisspl) [reply source/Irrlicht/CMeshManipulator.cpp]: But how you know that _inbuffer has a non null one?

File: include/CFinalBoneHierarchy.h [hunk 5]
```diff
@@ -3,14 +3,14 @@
 
 #include "assert.h"
 #include <vector>
-#include <unordered_set>
```
Comment (devshgraphicsprogramming) [inline include/CFinalBoneHierarchy.h:6]: You took away my unordered_set optimization it seems

Edit: checked the rest, it seems you didn't but you changed my include files :)

File: include/CFinalBoneHierarchy.h [hunk 6]
```diff
@@ -3,14 +3,14 @@
 
 #include "assert.h"
 #include <vector>
-#include <unordered_set>
+#include <set>
 #include <algorithm>
-#include <functional>
```
Comment (devshgraphicsprogramming) [inline include/CFinalBoneHierarchy.h]: functional is needed for delete and add keyframes functions

File: include/IMeshBuffer.h [hunk 7]
```diff
@@ -156,10 +158,387 @@ GL_INVALID_OPERATION is generated by glVertexAttribPointer if size is GL_BGRA an
...
+		}
+		return ECT_COUNT;
+	}
+
+	//! Gets not normalized, corresponding to given attribute type. Makes sense only for normalized types.
+	/**
+	@param _type Given type.
+	@returns Not normalized corresponding to given type or ECT_COUNT for types without such.
+	*/
+	inline E_COMPONENT_TYPE getCorrespondingNotNormalizedType(E_COMPONENT_TYPE _type)
```
Comment (devshgraphicsprogramming) [inline include/IMeshBuffer.h]: rename to denormalized for consistency

**Comment by devshgraphicsprogramming** [comment]
I have a feeling that 
https://github.com/Crisspl/IrrlichtBAW/blob/7c49843ece6f0f9ae170ffc990eee740b4d5027e/include/IMeshManipulator.h#L65
https://github.com/Crisspl/IrrlichtBAW/blob/7c49843ece6f0f9ae170ffc990eee740b4d5027e/include/IMeshManipulator.h#L128
Should disappear.

And also the `makePlanarTextureMapping` family of functions in CMeshManipulator.*

**Comment by devshgraphicsprogramming** [comment]
After my most recent comments are addressed I have 3 final questions before approving the merge:
1) Are the setAttribute and getAttribute tested for parity (same results) for the attribute types that used to work?? **
2) Is the matrixSIMD3x4 skinning upgrade well tested?
3) Have you tested that the mesh optimizer not only works, but that it actually reduces the mesh memory footprint (i.e. makes .baw files or vertex sizes smaller)???


** We have code like this in BAW
```c++
        size_t totalv=0;
        for (size_t n=0;n<tmpMesh->getMeshBufferCount();n++)
        {
            scene::ICPUMeshBuffer* meshbuffer = tmpMesh->getMeshBuffer(n);
            core::vectorSIMDf outPos(0.f,0.f,0.f,1.f);
            size_t v=0;
            for (; meshbuffer->getAttribute(outPos,meshbuffer->getPositionAttributeIx(),v); v++)
            {
                outPos *= sms->_scale_transformed;
                meshbuffer->setAttribute(outPos,meshbuffer->getPositionAttributeIx(),v);
            }
			totalv += v;
            core::aabbox3df box = meshbuffer->getBoundingBox();
            box.MinEdge *= sms->_scale_transformed;
            box.MaxEdge *= sms->_scale_transformed;
            meshbuffer->setBoundingBox(box);
        }
        core::aabbox3df box = tmpMesh->getBoundingBox();
        box.MinEdge *= sms->_scale_transformed;
        box.MaxEdge *= sms->_scale_transformed;
        tmpMesh->setBoundingBox(box);
```

The position attribute is always ECT_FLOAT with ECPA_THREE or ECPA_FOUR, if it doesn't work we will be unhappy.

**Comment by Crisspl** [comment]
> Is the matrixSIMD3x4 skinning upgrade well tested?
  
Hardware skinning example works fine (but faster :D), so... I think yes.

> Have you tested that the mesh optimizer not only works, but that it actually reduces the mesh memory footprint (i.e. makes .baw files or vertex sizes smaller)???
  
Yes, but usually it's not a lot. Also it's easly "lookable" with `-info` option of `convert2baw`.

> Are the setAttribute and getAttribute tested for parity (same results) for the attribute types that used to work??

Actually yes, I did that while debugging some time ago.

Btw, you can have multi-line code in this fine-looking code-block-thing by encapsulating code with triple thing-below-tilde, so it looks like this:
```c++
size_t totalv=0; 
for (size_t n=0;n<tmpMesh->getMeshBufferCount();n++) 
{
   scene::ICPUMeshBuffer* meshbuffer = tmpMesh->getMeshBuffer(n);
   core::vectorSIMDf outPos(0.f,0.f,0.f,1.f);
   size_t v=0;
   for (; meshbuffer->getAttribute(outPos,meshbuffer->getPositionAttributeIx(),v); v++)
   {
      outPos *= sms->_scale_transformed;
      meshbuffer->setAttribute(outPos,meshbuffer->getPositionAttributeIx(),v);
   } 
   totalv += v;
   core::aabbox3df box = meshbuffer->getBoundingBox();
   box.MinEdge *= sms->_scale_transformed;
   box.MaxEdge *= sms->_scale_transformed;
   meshbuffer->setBoundingBox(box); 
}
core::aabbox3df box = tmpMesh->getBoundingBox();
box.MinEdge *= sms->_scale_transformed;
box.MaxEdge *= sms->_scale_transformed;
tmpMesh->setBoundingBox(box);
```

**Comment by devshgraphicsprogramming** [comment]
can you merge with the latest commit (take care when merging IMeshBuffer.h) 
and then test that example 1,3-19 and 23-24 produce expected results?

**Pull request merged**